storage: TestReplicaTryAbandon failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/5bbaeed18ff46f64d4702c517d3e459bc6d3e9bd

Parameters:
```
TAGS=
GOFLAGS=-race
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=811643&tab=buildLog

```

I180802 05:55:06.273074 37548 storage/store.go:1496  [s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I180802 05:55:09.386359 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:10.386873 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:11.387320 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:12.387944 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:13.388431 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:14.388929 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:15.389410 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:16.389953 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:17.390482 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:18.390997 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:19.393206 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:20.393766 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:21.394276 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:22.394797 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:23.395292 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:24.395830 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:25.396311 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:26.396778 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:27.397254 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:28.397761 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:29.398260 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:30.398824 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:31.399334 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:32.399853 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:33.400337 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:34.400874 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:35.401895 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:36.402444 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:37.402967 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:38.403473 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:39.404011 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:40.404456 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:41.404980 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:42.405460 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:43.407258 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:44.407818 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:45.408323 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:46.408768 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:47.409244 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:48.409737 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:49.410208 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
I180802 05:55:50.410755 37548 storage/replica_test.go:7562  SucceedsSoon: expected empty command queue, but found
  1 0.000000123,30 ["acdfg")
panic: test timed out after 30m0s

goroutine 37761 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/go/src/testing/testing.go:1240 +0x146
created by time.goFunc
	/usr/local/go/src/time/sleep.go:172 +0x52

goroutine 1 [chan receive, 28 minutes]:
testing.(*T).Run(0xc420227590, 0x31cfadb, 0x15, 0x32f1020, 0xc4202af901)
	/usr/local/go/src/testing/testing.go:825 +0x597
testing.runTests.func1(0xc420227590)
	/usr/local/go/src/testing/testing.go:1063 +0xa5
testing.tRunner(0xc420227590, 0xc4202afb20)
	/usr/local/go/src/testing/testing.go:777 +0x16e
testing.runTests(0xc4203bab80, 0x49c7fc0, 0x204, 0x204, 0x4a)
	/usr/local/go/src/testing/testing.go:1061 +0x4e2
testing.(*M).Run(0xc4202ee380, 0x0)
	/usr/local/go/src/testing/testing.go:978 +0x2ce
github.com/cockroachdb/cockroach/pkg/storage_test.TestMain(0xc4202ee380)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/main_test.go:57 +0x2a9
main.main()
	_testmain.go:1088 +0x22b

goroutine 6 [syscall, 30 minutes]:
os/signal.signal_recv(0x89fc91)
	/usr/local/go/src/runtime/sigqueue.go:139 +0xa6
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x30
created by os/signal.init.0
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x4f

goroutine 26 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.flushDaemon()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:1171 +0x132
created by github.com/cockroachdb/cockroach/pkg/util/log.init.0
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:590 +0x189

goroutine 27 [chan receive, 30 minutes]:
github.com/cockroachdb/cockroach/pkg/util/log.signalFlusher()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:597 +0xf2
created by github.com/cockroachdb/cockroach/pkg/util/log.init.0
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:591 +0x1a1

goroutine 34 [select, 30 minutes, locked to thread]:
runtime.gopark(0x32f43e0, 0x0, 0x31b5f8b, 0x6, 0x18, 0x1)
	/usr/local/go/src/runtime/proc.go:291 +0xf9
runtime.selectgo(0xc420091f50, 0xc4200b8300)
	/usr/local/go/src/runtime/select.go:392 +0x11d4
runtime.ensureSigM.func1()
	/usr/local/go/src/runtime/signal_unix.go:549 +0x19f
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2361 +0x1

goroutine 37559 [semacquire, 28 minutes]:
sync.runtime_Semacquire(0xc4208948f4)
	/usr/local/go/src/runtime/sema.go:56 +0x39
sync.(*WaitGroup).Wait(0xc4208948e8)
	/usr/local/go/src/sync/waitgroup.go:129 +0xb3
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Wait(0xc420894870, 0x39d3c20, 0xc420651ec0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:172 +0x3f
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Wait-fm(0x39d3c20, 0xc420651ec0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3830 +0x56
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc420e5c090, 0xc420894090, 0xc420e5c080)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:199 +0x14b
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:192 +0xbb

goroutine 37548 [semacquire, 28 minutes]:
sync.runtime_Semacquire(0xc4208940bc)
	/usr/local/go/src/runtime/sema.go:56 +0x39
sync.(*WaitGroup).Wait(0xc4208940b0)
	/usr/local/go/src/sync/waitgroup.go:129 +0xb3
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc420894090, 0x39d3ba0, 0xc420052150)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x3ac
runtime.Goexit()
	/usr/local/go/src/runtime/panic.go:377 +0x120
testing.(*common).FailNow(0xc42186c000)
	/usr/local/go/src/testing/testing.go:553 +0x59
testing.(*common).Fatalf(0xc42186c000, 0x3212e79, 0x2d, 0xc420db2ae0, 0x3, 0x3)
	/usr/local/go/src/testing/testing.go:597 +0x95
github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3a07be0, 0xc42186c000, 0xc420e5c210)
	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:49 +0x2fc
github.com/cockroachdb/cockroach/pkg/storage.TestReplicaTryAbandon(0xc42186c000)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_test.go:7562 +0x91f
testing.tRunner(0xc42186c000, 0x32f1020)
	/usr/local/go/src/testing/testing.go:777 +0x16e
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:824 +0x565

goroutine 37552 [semacquire, 28 minutes]:
sync.runtime_SemacquireMutex(0xc42019dcfc, 0x900000000)
	/usr/local/go/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc42019dcf8)
	/usr/local/go/src/sync/mutex.go:134 +0x172
sync.(*RWMutex).Lock(0xc42019dcf8)
	/usr/local/go/src/sync/rwmutex.go:93 +0x49
github.com/cockroachdb/cockroach/pkg/util/syncutil.(*RWMutex).Lock(0xc42019dcf8)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/mutex_sync.go:68 +0x3b
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReadyRaftMuLocked(0xc42019dc00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:4088 +0x813
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReady(0xc42019dc00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:3844 +0xec
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processReady(0xc420560000, 0x39d3c20, 0xc420cec330, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3752 +0x104
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc420894870, 0x39d3c20, 0xc420cec330)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:215 +0x344
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x39d3c20, 0xc420cec330)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:166 +0x56
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc42190e520, 0xc420894090, 0xc42190e510)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:199 +0x14b
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:192 +0xbb

goroutine 37760 [chan send, 28 minutes]:
github.com/cockroachdb/cockroach/pkg/storage.TestReplicaTryAbandon.func1.1(0xc420b96d80, 0xc420bd8240, 0xc420b97140, 0xc4205519e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_test.go:7512 +0xce
created by github.com/cockroachdb/cockroach/pkg/storage.TestReplicaTryAbandon.func1
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_test.go:7509 +0x9d

goroutine 37550 [semacquire, 28 minutes]:
sync.runtime_notifyListWait(0xc4205c4fd0, 0x5)
	/usr/local/go/src/runtime/sema.go:510 +0x11a
sync.(*Cond).Wait(0xc4205c4fc0)
	/usr/local/go/src/sync/cond.go:56 +0x8e
github.com/cockroachdb/cockroach/pkg/storage/engine.(*RocksDB).syncLoop(0xc4205c4ea0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:654 +0x86
created by github.com/cockroachdb/cockroach/pkg/storage/engine.(*RocksDB).open
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:642 +0x607

goroutine 37762 [semacquire, 28 minutes]:
sync.runtime_Semacquire(0xc42019dd04)
	/usr/local/go/src/runtime/sema.go:56 +0x39
sync.(*RWMutex).RLock(0xc42019dcf8)
	/usr/local/go/src/sync/rwmutex.go:50 +0x95
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).leaseGoodToGo(0xc42019dc00, 0x39d3c20, 0xc420814a80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1305 +0x104
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).redirectOnOrAcquireLease(0xc42019dc00, 0x39d3c20, 0xc420814a80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1340 +0xd6
github.com/cockroachdb/cockroach/pkg/storage.(*Store).startLeaseRenewer.func1(0x39d3c20, 0xc420b84b10)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:1640 +0x215
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc4203221c0, 0xc420894090, 0xc420322160)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:199 +0x14b
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:192 +0xbb


ERROR: exit status 2

27 runs completed, 1 failures, over 42m50s
context canceled

```

--- a/pkg/storage/replica_test.go
+++ b/pkg/storage/replica_test.go
@@ -7503,8 +7503,18 @@ func TestReplicaTryAbandon(t *testing.T) {
 	proposalErrCh := make(chan error)
 
 	// Cancel the request before it is proposed to Raft.
+	var proposed int32
 	tc.repl.mu.Lock()
 	tc.repl.mu.submitProposalFn = func(result *ProposalData) error {
+		if atomic.AddInt32(&proposed, 1) != 1 {
+			// We only need to propose the command once for this test. Worse, a
+			// second proposal will cause the test to deadlock. How do we get a
+			// second proposal? reasonNewLeaderOrConfigChange ->
+			// refreshProposalsLocked(). Note that this form of refresh cannot
+			// currently be disabled via a testing knob.
+			return nil
+		}
+
 		cancel()
 		go func() {
 			<-proposalCh

